GTC-3144 Create carbon emissions titiler endpoint

danscales opened the pull request:
GTC-3144 Create carbon emissions titiler endpoint

This is using the colors/symbology from the existing GFW raster tile cache (gfw_forest_carbon_gross_emissions/v20240402)

It includes one filter which is tree cover density, using umd_tree_cover_density_2010/v1.6. This actually needs to be the 2000 version, I will change that before checkin, after I generate the proper COGs.

I added an option in reader.py to specify the default_band, since it's possible that the default band is not called default.tif.

All rendering is true_color (not encoded), so there is no render_type parameter.

Tested that the code runs successfully and delivers tile data, but haven't been able to look at the results visually, because of problems testing in dev.  Will test visually in staging.

danscales committed:
GTC-3144 Create carbon emissions titiler endpoint

This is using the colors/symbology from the existing GFW raster tile
cache.

It includes one filter which is tree cover density, using
umd_tree_cover_density_2010/v1.6. This actually needs to be the 2000
version, I will change that before checkin, after I generate the proper
COGs.

I added an option in reader.py to specify the default_band, since it's
possible that the default band is not called default.tif.

All rendering is true_color (not encoded), so there is no render_type
parameter.

jterry64 reviewed (approved):
@solomon-negusse @danscales code looks good, but I notice we don't usually add many tests in this repo. Do we have anyway to add unit tests for the TiTiler code? 

I imagine we can just separate out the functions and pass in ImageData directly. If it's something we still need to figure out, maybe we can do when we refactor things later.

solomon-negusse commented:
> @solomon-negusse @danscales code looks good, but I notice we don't usually add many tests in this repo. Do we have anyway to add unit tests for the TiTiler code?
> 
> I imagine we can just separate out the functions and pass in ImageData directly. If it's something we still need to figure out, maybe we can do when we refactor things later.

@jterry64 we do just that in https://github.com/wri/gfw-tile-cache/blob/production/tests/titiler/test_alerts_algo.py. Always room for more tests though

solomon-negusse commented on app/routes/titiler/gfw_forest_carbon_gross_emissions.py:
```diff
@@ -0,0 +1,80 @@
+    *,
+    version: GfwForestCarbonGrossEmissions,
+    xyz: Tuple[int, int, int] = Depends(raster_xyz),
+    tree_cover_density_threshold: Optional[int] = Query(
```
Would this be the place to limit the choices to the predefined thresholds (10, 30, etc.) or did we decide to leave it open?

danscales replied:
OK, talked to David Gibbs, and he said anything >= 30% tcd is fine.  So, I change the constraint to 'ge=30' as you mentioned.  Flagship front-end can constraint to 30/50/70 (or whatever) is they want.

solomon-negusse reviewed (approved):
Looks good to me besides the one question I left as comment.

solomon-negusse commented on app/routes/titiler/gfw_forest_carbon_gross_emissions.py:
```diff
@@ -0,0 +1,80 @@
+    tile_x, tile_y, zoom = xyz
+    bands = ["tcd_30", "intensity"]
+    folder: str = f"s3://{DATA_LAKE_BUCKET}/{dataset}/{version}/raster/epsg-4326/cog"
+    with AlertsReader(input=folder, default_band="tcd_30") as reader:
```
Could the `tcd_30` name mean that this carbon emissions band is only applicable to TCD threshold of 30% and that we'd have to read a different carbon emissions band based on what threshold users input using the `tree_cover_density_threshold` query parameter?

danscales replied:
Yes, I had originally did a COG on the raster that had 30% baked in.  That may still work for >=30% filters, since any greater canopy would filter out even more.  But I've now generated a default.tif COG based on the base raster (no canopy filter baked in), as you are mentioning, and it now references a default.tif COG.

danscales committed:
Switch to using umd_tree_cover_density_2000 and the base emissions tif

For emissions, I should be using the umd_tree_cover_density_2000 as the
tcd filter. And I originally created COG from the Mg_CO2e_ha-1__tcd_30
raster of gfw_forest_carbon_gross_emissions, but I really wanted the
default asset Mg_CO2e_ha-1, from which I created a default.tif COG.

danscales committed:
Add constraint that tcd threshold should be >= 30

solomon-negusse committed:
use predefined TCD filters for emissions
this will be faster for clients as we pregenerate the emissions layer with the filter baked in

solomon-negusse commented:
@danscales: I have updated the PR limiting TCD filter to the predefined thresholds (David confirmed there isn't much need for arbitrary thresholds) and using emissions layers with that baked in. This will result in better performance loading the tiles as we don't have to read an extra filter layer and the emissions layer itself is smaller too with filtered pixels being masked out. Of course we need to pregenerate a few more COGs for this but this dataset is updated annually so the additional cost is not significant.

solomon-negusse committed:
use default emissions layer with correct tcd threshold

danscales commented on app/routes/titiler/gfw_forest_carbon_gross_emissions.py:
```diff
@@ -0,0 +1,75 @@
+    ),
+    tree_cover_density_threshold: Optional[TreeCoverDensityThreshold] = Query(
+        TreeCoverDensityThreshold.tcd_30,
+        description="Show alerts in pixels with tree cover density (in percent) greater than or equal to this threshold. `umd_tree_cover_density_2010` is used for this masking.",
```
I noticed I didn't change this comment, it should be `umd_tree_cover_density_2000`   Plus I guess you could mention that the COGs used also already filtered by mangroves and tree cover height, or you can leave out if you think that's too specific.

danscales reviewed (commented):
@solomon-negusse Looks great, except for my one minor comment.   I can't approve, since I created the PR, but you guys have already approved.

solomon-negusse committed:
fix tree cover density dataset version
